nntrainer/layers/conv2d_layer.cpp

```diff
@@ -26,30 +26,123 @@
+ * CONV2D_DIM> &dilation)
+ */
+template <>
+void conv2d_op<ChannelToCol>(const Tensor &in, const Tensor &filter,
+                             Tensor &out,
+                             const std::array<unsigned, CONV2D_DIM> &padding,
+                             const std::array<unsigned, CONV2D_DIM> &stride,
+                             const std::array<unsigned, CONV2D_DIM> &dilation) {
+
+  NNTR_THROW_IF(in.channel() != filter.channel(), std::invalid_argument)
+    << "channel mismatch, in_dim: " << in.getDim()
+    << "out_dim: " << out.getDim();
+
+  TensorDim filter_dim = filter.getDim();
+  Tensor f = filter;
+
+  f.reshape({filter.batch(), filter.getDim().getFeatureLen()});
+
+  // todo: allocate im2col_result beforehand to save time
+  Tensor im2col_result;
+  for (unsigned int b = 0; b < in.batch(); ++b) {
+    Tensor out_sub = out.getBatchSlice(b, 1);
+    Tensor in_sub = in.getBatchSlice(b, 1);
+
+    out_sub.reshape({out.channel(), out.width() * out.height()});
+    Conv2DLayer::im2col(in_sub, filter_dim, padding, stride, true,
+                        im2col_result);
+    f.dot(im2col_result, out_sub, false, true);
+  }
+}
+
+/**
+ * @brief implementation of convolution ops needed when channel need to go to
+ * row, the example describes what to expect
+ * eg)    x     \*\       dy        =       dw, where \*\ denotes conv operation
+ * (N, C, H, W) \*\ (N, FN, OH, OW) = (FN, C, FH, FW)
+ * ==> for each batch, slice dy, x
+ * [1] imx <- im2col(x)
+ * [2]   dy      x        imx             =         dw
+ * (FN, OH * OW) x (OH * OW, C * FH * FW) = (FN, C * FH * FW)
+ * [3] dw <- reshape(dw, (FN, C, FH, FW))
+ *
+ * @copydoc template<typename T>conv2d_op(const Tensor &in, const Tensor
+ * &filter, Tensor &out, const std::array<unsigned, CONV2D_DIM> &padding, const
+ * std::array<unsigned, CONV2D_DIM> &stride, const std::array<unsgined,
+ * CONV2D_DIM> &dilation)
+ */
+template <>
+void conv2d_op<ChannelToRow>(const Tensor &in, const Tensor &filter,
+                             Tensor &out,
+                             const std::array<unsigned, CONV2D_DIM> &padding,
+                             const std::array<unsigned, CONV2D_DIM> &stride,
+                             const std::array<unsigned, CONV2D_DIM> &dilation) {
+  Tensor f = filter;
+  f.reshape(
+    {filter.batch(), 1, filter.channel(), filter.height() * filter.width()});
+
+  Tensor o = out;
+  o.reshape({out.batch(), out.channel() * out.height() * out.width()});
+  o.setZero(); // initialize to 0 as it accumulates
+
+  // todo: allocate im2col_result beforehand to save time
+  Tensor im2col_result;
+  for (unsigned int b = 0; b < in.batch(); ++b) {
+    Tensor in_sub = in.getBatchSlice(b, 1);
+    Tensor filter_sub = f.getBatchSlice(b, 1);
+
+    Conv2DLayer::im2col(in_sub, filter.getDim(), padding, stride, false,
+                        im2col_result);
+
+    filter_sub.dot(im2col_result, o, false, false, 1.0f);
```

Comment:
How about pass beta as 0.0f if b == 0 and 1.0f when b > 0
eliminating the need for o.setZero() at line 130?

Reply:
Done